### ISSUE TITLE
CommentAuthor should have RootQuery entry points in the Graph

### ISSUE THREAD

**Issue by jasonbahl** [wp-graphql/wp-graphql#1757]
The `CommentAuthor` is a Type in the graph that is used to represent authors of comments that are not users. 

When querying comments and their authors, you can query like so: 

```
{
  comments {
    nodes {
      id
      content
      author {
         node {
           __typename
           id
          ...on User {
            firstName
            lastName
          }
          ...on CommentAuthor {
             name
          }
         }
      }
    }
  }
}
```

Currently, the only access point to get to an object of the `CommentAuthor` type is to fetch it as a relationship of a comment. There's no direct Root entry points in the Graph to get a `CommentAuthor`.

But, since the `CommentAuthor` implements the `Node` Interface, it should have Root entry points into the Graph. 

I propose adding a `commentAuthors { ... }` Root entry point as well as a singular `commentAuthor( id: $id ) { ... }` entry point to the graph. 

----

Related: https://github.com/gatsbyjs/gatsby/issues/29998

**Comment by moonmeister** [comment]
Decided to give this a shot. I'm making some progress but need some clarification. Currently when querying cor comment authors on a comment, if the commenter wasn't a logged in user, the `CommentAuthor=>databaseId` will equal the `Comment=>databaseId` but the global `id` appears to be unique. Additionally if the comment user is logged in, the database ID becomes the `User=>databaseId` and not the `Comment=>databaseId` , Does this mean we are only allow the `commentAuthor( id: $id ) { ... }` entry point to use Global IDs and not database IDs? Or is this a bug that needs smoothing out?

**Comment by justlevine** [comment]
@moonmeister great question!

I _think_ we can support `idType: DATABASE_ID` with some conditional logic, since database ids are unique in the WP backend. I.e. check if the ID points to a valid user, and if not then use the comment to resolve instead (or vice versa).

**Comment by jasonbahl** [comment]
ok, so we have a few things to consider here. 

`CommentAuthor` is treated as an author that's not a user. 

So, querying a `CommentAuthor` by ID should use the `comment->databaseId` (see: https://github.com/wp-graphql/wp-graphql/blob/develop/src/Model/CommentAuthor.php#L51-L56)

We should be able to query like so: 

```graphql
{
  commentAuthor( id: 123 idType: DATABASE_ID ) { ... }
  commentAuthor( id: "comment_author:123 ) { ... } # would be base_64_encoded Id, of course
}
```

And under the hood that should query for the comment with databaseId `123`, then use the comment object to return a [CommentAuthor Model](https://github.com/wp-graphql/wp-graphql/blob/develop/src/Model/CommentAuthor.php#L36).

----

Now, if we wanted to be able to query for a `commenter` from the root, things would be different, as `commenter` is an Interface that can be a user or a comment_author, and in that case, `commenter` does not have unique IDs, as you could have a comment (and therefore CommentAuthor) with database id: `123` in the comments table, and also have a user (commenter) with id `123` in the user table.

---

so, querying for `commentAuthor()` yes, we should be able to support databaseId. 

Querying for `commentor` (if we chose to add that as an entry point) would not. It would need to be something like: 

```graphql
{
  commenter( id: "user:123" ) { ... }
  commenter( id: "comment_author:123" ) { ... }
  commenter( id: 123 idType: USER_DATABASE_ID ) { ... }
  commenter( id: 123 idType: COMMENT_AUTHOR_DATABASE_ID { ... }
}
```

**Comment by moonmeister** [comment]
> CommentAuthor is treated as an author that's not a user.
In theory yes, returning `__typname` shows this when resolving `Comment=>author`.
In practice, when using `DataSource::resolve_comment_author()` It will return a `User` as a `CommentAuthor`. This may be bug?

As far as `CommentAuthor` vs `Commenter`...Is there some background or examples on what inspired this feature that would help determine the need? What are the use cases for either or both?

**Comment by jasonbahl** [comment]
> This may be bug?

yes. Sounds like a bug. The `resolve_comment_author` method should probably be changed to require a type as well so it knows to load from the comment table or user table

**Comment by jasonbahl** [comment]
>As far as CommentAuthor vs Commenter...Is there some background or examples on what inspired this feature that would help determine the need? What are the use cases for either or both?

@moonmeister WordPress core functionality lets you show different data depending on whether the Author is a user or not. 

And being a "graph" we expose that to users to consume as well. 

If the `Commenter` is a User, you can enter the graph and get user data. 

If the `Commenter` is not a user, you're limited in the information you can ask for. 

For example, I can do something like: 

```
{ 
  comment( id: 123 idType: DATABASE_ID ) {
    content
    author {
       id
       name
       ...on User {
         recentPosts: posts {
           nodes { 
              id
              title
            }
          }
       }
    }
  }
}
```

If the Author of the comment is a `User`, I could expose additional data about them, like recent posts they've authored, and could add it to the UI or whatever. 

All sorts of use cases here!

**Comment by moonmeister** [comment]
Right, that makes sense. I guess the question is this top level entry point only include `CommentAuthors` or does it include `Users` as well? What' sthe use case not of the individual types, but of the RootQuery entry point?

**Comment by jasonbahl** [comment]
I think we have room to add both. 

An entry point to get `CommentAuthor` by ID from the root (these would _not_ be users)

Another entry point to get `Commenter` by ID from the root (these could be _users_ or "guest" authors / CommentAuthor type).

The original ask was for the `commentAuthor` entry point, so let's start with that?

And we can work on `commenter` entry separately

Think of it like the different entry points for posts. 

You could do: 

```
{ post( id: "..." ) { ... } }

or

{ contentNode( id: "..." ) { ...on Post { ... } } }
```

Multiple ways to enter the graph!

**Comment by jasonbahl** [comment]
> is there a scenario where I explicitly wouldn't want Users?

Sure! 

What if i had a list of comments that showed the author, and when I click the author a Modal or Drawer opens with information about the author. 

If the Author is a `User` a `UserCommentor` component could be returned which queried data about the user to show in the Modal/Drawer.

If the Author is a `CommentAuthor` (guest author) a different component `CommentAuthorBio` or whatever, could be returned which queries for data available to that type. 

psuedo code: 

```js

const UserCommentorBio = ( { databaseId } ) => { 
  const { data } = useQuery( `
    query GetUserBio( $id: ID! ) {
       user( id: $id idType: DATABASE_ID ) {
          id
          name
          recentPosts: posts { nodes { id, title, date } }
       }
    }
  `, { variables: { id: databaseId } }  )
  
  return (
    <div>
      <pre>{JSON.stringify( data, null, 2 )</pre>
    </div>
  )
}

const CommentAuthorBio = ( { databaseId } ) => {
  const { data } = useQuery( `
    query GetCommentAuthorBio( $id: ID! ) {
       commentAuthor( id: $id idType: DATABASE_ID ) {
          id
          name
       }
    }
  ` , { variables: { id: databaseId } } )
  
  return (
    <div>
      <pre>{JSON.stringify( data, null, 2 )</pre>
    </div>
  )
}

const Comment = ({ databaseId }) => {
  
  const { data } = useQuery( `
  query GetComment( $id: ID! ) {
    comment( id: $id ) { 
      content
      author {
        node {
          __typename
          databaseId
        }
      }
    }
  }
  `, { variables: { id: databaseId } } );
  
  const type = data?.comment?.author?.node?.__typename;
  
  if ( type === 'CommentAuthor ) { 
    return <CommentAuthorBio databaseId={ data?.comment?.author?.node?.databaseId } />
  }
  
   if ( type === 'User ) { 
    return <UserCommentorBio databaseId={ data?.comment?.author?.node?.databaseId } />
  }

 return null;
}

```

**Comment by jasonbahl** [comment]
^ very contrived, pseudo code, but hope it conveys roughly a use case
